src/frontend/src/binder/mod.rs

```diff
@@ -54,9 +54,18 @@ pub struct Binder {
     db_name: String,
     context: BindContext,
     /// A stack holding contexts of outer queries when binding a subquery.
+    /// It also holds all of the stashed table contexts for each respective
+    /// subquery.
     ///
     /// See [`Binder::bind_subquery_expr`] for details.
-    upper_contexts: Vec<BindContext>,
+    upper_subquery_contexts: Vec<(BindContext, Vec<BindContext>)>,
+
+    /// A stack holding stashed contexts of non-child `TableFactor`s when binding
+    /// column expressions in a join.
+    ///
+    /// We need a separate stack as `CorrelatedInputRef` depth is
+    /// determined by the upper subquery context depth, not the table context stack depth.
+    upper_table_contexts: Vec<BindContext>,
```

Comment:
Is it necessary to distinguish them? This looks very like a tree. I guess by putting ctx at proper levels can also avoid incorrect subquery depth? 
```
subquery ctx
    |        \          \
subquery ctx   table ctx1  table ctx2 ...
    |        \
subquery ctx   ...
```
BTW, I'm also not sure whether this solution works in really complicated cases, e.g., a table factor is itself a subquery which has a correlated subquery inside? 🤔

Reply:
And I think a proper fix can revert the workaround in https://github.com/singularity-data/risingwave/pull/3352? 🤔

Reply:
I think this fix does indeed agree with this comment: https://github.com/singularity-data/risingwave/pull/3352#issuecomment-1161176436

Left table references are invalid, but outer query references are valid is correlated input refs.

Reply:
> subquery which has a correlated subquery inside? 🤔

In the case of nested subqueries, outer queries will have their local tables stashed in the subquery_contexts stack. You can see the logic in bind_columns, we check for correlated input refs in all contexts local to upper subqueries (but not any tables to the left within the subquery).

Reply:
I think proper levels is not enough. When we have nested joins, then we need to be able to push the table context at the current subquery level multiple times.

Reply:
I think lateral join will be handled adequately by allowing correlated input refs to be bound from the left table iff the subquery is a lateral join.

Reply:
> And I think a proper fix can revert the workaround in https://github.com/singularity-data/risingwave/pull/3352? thinking

I believe this fix will not completely go away. We still may not want correlated subqueries not marked as lateral that are not in the `WHERE` clause. Need some clarification on how powerful exactly is arbitrary query unnesting (some examples of what it should be able to do would be nice).

Btw, I think we should also be cautious about relying too much on postgres. Postgres optimizer is known to be limited, especially when dealing with subqueries.

Reply:
> BTW, I'm also not sure whether this solution works in really complicated cases, e.g., a table factor is itself a subquery which has a correlated subquery inside? 

I've added test cases for this